clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -4684,6 +4684,29 @@ LValue CodeGenFunction::EmitLValueForLambdaField(const FieldDecl *Field,
     else
       LambdaLV = MakeAddrLValue(AddrOfExplicitObject,
                                 D->getType().getNonReferenceType());
+
+    // Make sure we have an lvalue to the lambda itself and not a derived class.
+    auto *ThisTy = D->getType().getNonReferenceType()->getAsCXXRecordDecl();
+    auto *LambdaTy = cast<CXXRecordDecl>(Field->getParent());
+    if (ThisTy != LambdaTy) {
+      CXXBasePaths Paths(/*FindAmbiguities=*/false, /*RecordPaths=*/true,
+                         /*DetectVirtual=*/false);
+
+      [[maybe_unused]] bool Derived = ThisTy->isDerivedFrom(LambdaTy, Paths);
+      assert(Derived && "Type not derived from lambda type?");
```

Comment:
Is there a reason there has to be a unique base subobject of the lambda type?

Reply:
That’s something I’m not entirely sure about, but I couldn’t come up with a way to actually end up with more than one path from a derived class to the lambda’s type such that both would still be accessible.

Reply:
We can manually resolve between the `operator()`s at the call site with a qualified member name, and the `operator()` should still get called with the original `Overloaded` without any hint about which subobject it was called on.

We could also have `Overloaded` just pick one of its bases to `using` the `operator()` from, but then I suppose someone might argue that the operator ought to pick up on that to decide the base path.  (That would be absurd, IMO.)

Example of the first: https://godbolt.org/z/rqvKGfevY

I think you need to open a DR with the committee.  It just needs to be ill-formed for a lambda call operator with an explicit `this` parameter to be instantiated such that the `this` parameter has a type with multiple base subobjects of the lambda type.  I can't imagine any other resolution short of something absurd like adding an implicit offset parameter, which would then be breaking in other ways.

Please go ahead and diagnose that condition in Sema.

Reply:
> We can manually resolve between the `operator()`s at the call site with a qualified member name

Ah, that’s what I was missing.

> I think you need to open a DR with the committee.

That sounds reasonable, but I candidly have no idea how to do that.

> Please go ahead and diagnose that condition in Sema.

Will do.

Reply:
https://isocpp.org/std/submit-issue for defect reports

Also, maybe we should record the correct path in the AST somewhere?

Reply:
> Also, maybe we should record the correct path in the AST somewhere?

That would probably be less of a hack than doing this here, but the main problem I see is that all we have here is a DRE; it just seems a bit strange to store cast paths in one—at least to me.

Reply:
Richard got you beat by a few days https://cplusplus.github.io/CWG/issues/2881.html

Reply:
If we were going to put it in the AST, the right place would be on the call operator FD rather than in every DRE for a capture.  We'd presumably store it with a side-table in the ASTContext, which we'd only populate for lambdas with a derived object parameter.  It's probably not an actual problem to repeatedly recompute it as long as we have that restriction, though.

Reply:
> We'd presumably store it with a side-table in the ASTContext, which we'd only populate for lambdas with a derived object parameter.

I was just thinking about storing it in the call operator but I wasn’t sure where, but this makes sense.

Reply:
On that note, as for actually emitting the cast, is it worth it doing that once when we start emitting the lambda and then just reusing the computed lvalue instead of emitting the cast every time we need to access a capture? Or do we just let the optimiser clean that up? Asking because I’m not too familiar w/ how situations like these are typically handled in codegen.

Reply:
We could do either.  Doing it once in the prologue makes some sense to me; in the (very unlikely) case that it's not just a static offset, that would probably lead to better code.